### PR TITLE
[DICT] Add Python support for dict() function

### LINKED ISSUE #1894: Add Python support for dict() function
### Describe the feature
I would like to be able to use the `dict()` function. The `dict()` function is different to the `{}` function as it can behave as a copy function and/or create a dictionary using keyword arguments.

### Test Code
Provide code which does not currently work but which should do when this issue is fixed:
```python
if __name__ == '__main__':
    a = dict({1:1.0,2:2.0})
    b = dict(a=1, b=2)
```

### Proposed Solution
This feature should be implemented similarly to what was done for tuples. The equivalent issues for tuples was resolved in #1589.
The pertinent code is:
https://github.com/pyccel/pyccel/blob/d64be7ff7072faa07ce19c9e70fa02ed45a25c9c/pyccel/ast/builtins.py#L596-L620

### Additional context
It is important to handle the `dict` function so that this function can be used for type annotations. If the `{,}` function is used bugs such as #1892 can arise.

### PR DESCRIPTION
**Pull request by EmilyBourne**
Add Python support for `dict()` function as a copy operator or with key word arguments to build a dictionary. This fixes #1894 . The introduction of the `dict` function allows variable annotations to be supported. `__eq__` and `__hash__` functions are added to container types. This allows the initialisation of empty containers with syntax like `a : dict[str,int] = {}`.

### CODE REVIEW

File: pyccel/ast/builtins.py [hunk 1]
```diff
@@ -851,6 +852,58 @@
...
+        super().__init__(copied_obj)
+
+    @property
+    def copied_obj(self):
+        """
+        The object being copied.
+
+        The object being copied to create a new dict instance.
+        """
+        return self._args[0]
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/builtins.py:905]: /bot accept
Getter will be tested when low-level support for copies is added

**Review by mustapha-belbiad — APPROVED** [review]
Looks Good to Me.

**Review by jalalium — APPROVED** [review]

File: pyccel/ast/builtins.py [hunk 2]
```diff
@@ -882,6 +883,56 @@ def values(self):
...
+
 #==============================================================================
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/builtins.py:936]: /bot accept
Getter will be tested when low-level support for copies is added

**Review by jalalium — APPROVED** [review]

**Pull request merged**